### PR TITLE
Various improvements to the library

### PR DESCRIPTION
**Pull request by mulimoen**
Hi! I've experienced segfaults while running `cargo test`, and tried various things to narrow down this error. It turns out this was due to `blosc_get_complib_info` requiring the two last arguments to be non-Null as well as not supporting a `compname` not defined by the library.

To ensure such errors will not crop up again I've added some github actions, and ensured the bindings are always created at buildtime using the same configuration as the build.

### CODE REVIEW

**Comment by mulimoen** [comment]
I don't know what the best option is for testing with a downloaded `c-blosc`. I would like to see source as a separate package (`blosc-src`) which can be linked to `blosc-sys` using a feature.

File: blosc/Cargo.toml [hunk 1]
```diff
@@ -13,13 +13,9 @@ categories = ["api-bindings"]
 keywords = ["compression"]
 
 [dependencies]
-blosc-sys = "1.14.4"
+blosc-sys = { version = "1.14.5-pre", path = "../blosc-sys" }
```
Comment (asomers) [inline blosc/Cargo.toml:16]: The version number is based on whichever c-blosc library blosc-sys was built for.  If you dynamically generate bindings, then it's not possible to use this versioning scheme.

File: blosc/tests/test.rs [hunk 2]
```diff
@@ -10,11 +10,6 @@ use std::collections::hash_map::DefaultHasher;
 use std::hash::{Hash, Hasher};
 use std::mem;
 
-#[test]
-fn test_invalid_compressor() {
-    assert!(Context::new().compressor(Compressor::Invalid).is_err())
```
Comment (asomers) [inline blosc/tests/test.rs:14]: The purpose of this test is to achieve coverage of the C blosc_get_complib_info function, so you should leave it in.

File: blosc/tests/test.rs [hunk 3]
```diff
@@ -63,7 +68,11 @@ test_suite! {
         }).collect::<Vec<_>>();
 
 
-        let encoded = settings.val.compress(&sample[..]);
+        let encoder = settings.val;
+        if encoder.is_err() {
+            return
```
Comment (asomers) [inline blosc/tests/test.rs:73]: Why return?  That would pass the test.  Better to panic in this case.

File: blosc-sys/build.rs [hunk 4]
```diff
@@ -15,6 +15,18 @@ fn main() {
     } else {
         None
     };
+    #[cfg(unix)]
```
Comment (asomers) [inline blosc-sys/build.rs:18]: Why bother with `#[cfg(unix)]`?  Are you trying to use this crate on Windows?

File: blosc-sys/build.rs [hunk 5]
```diff
@@ -15,6 +15,18 @@ fn main() {
...
+    if bindings.is_none() {
+        if let Ok(pc) = pkg_config::probe_library("blosc") {
```
Comment (asomers) [inline blosc-sys/build.rs:20]: If you're going to go to the trouble of using pkg-config for this case when the header isn't in the default include path, why don't just always use pkg-config?

**Comment by mulimoen** [comment]
Thanks for the review! I'll split this PR up into more manageble chunks!

>     * I generally adhere to libc's philosophy that bindgen is an imperfect tool, and its output will in general need to be tweaked.  That's why I checked in the generated bindings.  Do you have a good reason why I shouldn't continue to do that?

It is a necessary to ensure the API does not change between versions, and bindgen is one way of doing this. If you have checked that there are no API changes for all published releases of `blosc` the bindings can be pregenerated and released.

File: blosc/Cargo.toml (hunk 1 above)
Comment (mulimoen) [reply blosc/Cargo.toml:16]: So `1.14 ` corresponds to `blosc` 1.14? How do you deal with breaking changes in the `rust` bindings but not `blosc`?

File: blosc/tests/test.rs (hunk 2 above)
Comment (mulimoen) [reply blosc/tests/test.rs:14]: This segfaults on some versions of `blosc`

File: blosc/tests/test.rs (hunk 3 above)
Comment (mulimoen) [reply blosc/tests/test.rs:73]: Not all versions of blosc ships with all decoders. Better to have a fail-open and ensure the CI uses all encoders.

File: blosc-sys/build.rs (hunk 4 above)
Comment (mulimoen) [reply blosc-sys/build.rs:18]: This is for detection through `pkg-config`. The crate should ideally be usable on `windows`.

File: blosc-sys/build.rs (hunk 5 above)
Comment (mulimoen) [reply blosc-sys/build.rs:20]: Good idea! I'll change this up! The idea was to have this as a fallback only for supported platforms.

**Comment by asomers** [comment]
> It is a necessary to ensure the API does not change between versions, and bindgen is one way of doing this. If you have checked that there are no API changes for all published releases of `blosc` the bindings can be pregenerated and released.

You mean no API changes between versions of c-blosc?  There had better not be any!  c-blosc publishes a shared library.  If there are any backwards-incompatible changes, then all programs will be break, whether they're written in Rust or C.  We can rely on c-blosc being backwards-compatible.  Forwards-compatibility, however, is not guaranteed.  That's one reason to pregenerate bindings.  If we generate bindings dynamically, then somebody's build server might bind a newer version of c-blosc than what they're using on their production machines, where the binary may not work.

**Comment by mulimoen** [comment]
You are raising some very good points here. I see on their release page they are very cautios about breaking the ABI.

It would be nice if one could specify which `blosc` to use, maybe through an enviroment variable?

**Comment by asomers** [comment]
> You are raising some very good points here. I see on their release page they are very cautios about breaking the ABI.
> 
> It would be nice if one could specify which `blosc` to use, maybe through an enviroment variable?

This can be done with crate features.  For example, if there were a cool feature in c-blosc 1.16.0, then the blosc could add a "1.14.4-compat" feature.  When enabled, it would only include functionality that was available as of version 1.14.4.  But when disabled, it would include newer functionality.

**Comment by mulimoen** [comment]
> This can be done with crate features. For example, if there were a cool feature in c-blosc 1.16.0, then the blosc could add a "1.14.4-compat" feature. When enabled, it would only include functionality that was available as of version 1.14.4. But when disabled, it would include newer functionality.

I was thinking at link time to emit the correct -L and -l for `.so` version compatability.

**Comment by asomers** [comment]
> I was thinking at link time to emit the correct -L and -l for `.so` version compatability.

I haven't figured out how to make Cargo link to a specific shlib version.  Do you know how?

**Comment by mulimoen** [comment]
The linker will decide based on whats available, but the `cargo:rustc-link-search=native=PATH` directive will at least add the path to the search path. I'm not what the order of preference is.

**Comment by mulimoen** [comment]
I really like the approach `hdf5-rust` has taken to versioning, where all features available at build time is included in the library.

I'm going to close this PR as bindgen at build time is not relevant.

**Pull request closed without merging**